File: src/main/java/sage/entity/Blog.java
package sage.entity;

import java.util.Date;
import java.util.HashSet;
import java.util.Set;
import javax.persistence.*;

import sage.domain.commons.IdCommons;

@Entity
public class Blog {
  private Long id;
  private String title;
  private String content;
  private User author;
  private Date createdTime;
  private Date modifiedTime;
  private Set<Tag> tags = new HashSet<>();

  Blog() {}

  public Blog(String title, String content, User author, Date createdTime, Set<Tag> tags) {
    this.title = title;
    this.content = content;
    this.author = author;
    this.createdTime = createdTime;
    this.tags.addAll(tags);
  }

  @Id
  @GeneratedValue
  public Long getId() {
    return id;
  }
  void setId(Long id) {
    this.id = id;
  }

  public String getTitle() {
    return title;
  }
  public void setTitle(String title) {
    this.title = title;
  }

  @Column(columnDefinition = "TEXT") @Lob
  public String getContent() {
    return content;
  }
  public void setContent(String content) {
    this.content = content;
  }

  @ManyToOne(optional = false)
  public User getAuthor() {
    return author;
  }
  void setAuthor(User author) {
    this.author = author;
  }

  public Date getCreatedTime() {
    return createdTime;
  }
  void setCreatedTime(Date createdTime) {
    this.createdTime = createdTime;
  }

  public Date getModifiedTime() {
    return modifiedTime;
  }
  public void setModifiedTime(Date modifiedTime) {
    this.modifiedTime = modifiedTime;
  }

  @ManyToMany(fetch = FetchType.EAGER)
  public Set<Tag> getTags() {
    return tags;
  }
  public void setTags(Set<Tag> tags) {
    this.tags = tags;
  }

  @Override
  public String toString() {
    return author + ": " + title + tags
        + "\n" + content;
  }

  @Override
  public int hashCode() {
    return IdCommons.hashCode(getId());
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    
    Blog other = (Blog) obj;
    return IdCommons.equal(getId(), other.getId());
  }
}

File: src/main/kotlin/sage/entity/Blog.kt
package sage.entity

import com.avaje.ebean.annotation.SoftDelete
import sage.domain.constraints.BlogConstraints
import sage.transfer.SearchableBlog
import java.sql.Timestamp
import java.util.*
import javax.persistence.*

@Entity
class Blog(
    var title: String,

    @Column(columnDefinition = "TEXT") @Basic(fetch = FetchType.LAZY)
    var inputContent: String,

    hyperContent: String,

    @ManyToOne(optional = false)
    val author: User,

    tags: Set<Tag>,

    var contentType: Short
) : AutoModel() {

  @Column(columnDefinition = "TEXT") @Basic(fetch = FetchType.LAZY)
  var content: String = hyperContent

  @ManyToMany
  var tags: MutableSet<Tag> = HashSet(tags)

  var whenEdited: Timestamp? = null

  @Column(nullable = false)
  var tweetId: Long = 0

  @SoftDelete
  var deleted: Boolean = false

  fun stat() = BlogStat.byId(id)

  fun validate() {
    BlogConstraints.check(this)
  }

  fun toSearchableBlog(): SearchableBlog = SearchableBlog(
      id, author.toUserLabel(), title, content, whenCreated, whenEdited, tags.map(Tag::toTagLabel)
  )

  companion object : BaseFind<Long, Blog>(Blog::class) {
    val MARKDOWN: Short = 1
    val RICHTEXT: Short = 2
    fun contentTypeValue(contentType: String): Short = when(contentType.toLowerCase()) {
      "markdown" -> MARKDOWN
      "richtext" -> RICHTEXT
      else -> 0
    }
    fun contentTypeString(contentType: Short): String = when(contentType) {
      MARKDOWN -> "markdown"
      RICHTEXT -> "richtext"
      else -> ""
    }

    fun byAuthor(authorId: Long): List<Blog> = where().eq("author", User.ref(authorId)).orderBy("id desc").findList()
  }
}